airflow/utils/helpers.py

```diff
@@ -224,47 +227,45 @@ def kill_process_tree(logger, pid, timeout=DEFAULT_TIME_TO_WAIT_AFTER_SIGTERM):
 
     # Check child processes to reduce cases where a child process died but
     # the PID got reused.
-    descendant_processes = [x for x in root_process.children(recursive=True)
-                            if x.is_running()]
+    running_descendant_processes = [x for x in root_process.children(recursive=True)
+                                    if x.is_running()]
 
-    if len(descendant_processes) != 0:
+    if len(running_descendant_processes) != 0:
         logger.info("Terminating descendant processes of {} PID: {}"
-                    .format(root_process.cmdline(),
-                            root_process.pid))
-        temp_processes = descendant_processes[:]
-        for descendant in temp_processes:
+                    .format(root_process.cmdline(), root_process.pid))
+        for descendant in running_descendant_processes:
             logger.info("Terminating descendant process {} PID: {}"
                         .format(descendant.cmdline(), descendant.pid))
-            if not kill_using_shell(logger, descendant.pid, signal.SIGTERM):
-                descendant_processes.remove(descendant)
+            kill_using_shell(logger, descendant.pid, signal.SIGTERM)
```

Comment:
What are the benefits of using the shell to do this? I only see downsides

Reply:
@bolkedebruin the alternative could be along the lines of what @milanvdm proposed in #2877. I think that's a more flexible/safe solution, assuming we remove this shell-based code all together. My only concern is having to do db migration on TaskInstance, which could be difficult for long running DBs.

Reply:
The other alternative to be `os.kill()`, or given we already have psutil.Process objects `descendant.terminate()` and then `descendant.kill()`

See also http://psutil.readthedocs.io/en/latest/#terminate-my-children

Reply:
@ashb  ah gotcha. I assume that's what @bolkedebruin meant. Sure let me take a look.